### PR TITLE
Introduce support for custom error codes

### LINKED ISSUE #275: “ledger_entry” with request improperly specified a Currency Code field do not return "malformedCurrency"
Expected response: `malformedCurrency`
Actual response: `invalidParams`

Sample request:
`{
  "method": "ledger_entry",
  "params": [{
    "ripple_state": {
      "accounts": [
        "rEgAvY1qyEfZxwSaDg5UdLtNQzyvSPEXtV",
        "rhyxmFfxujVTfxBjug37kouzms6YzWpAUy"
      ],
      "currency": "aa"
    },
    "ledger_index": "validated"
  }]
}`

### PR DESCRIPTION
**Pull request by godexsoft**
Fixes #275 

- Introduce a CombinedError that combines rippled errors with clio's custom ones
- Refactor a bit to make it more readable + add doxygen comments
- Add unit tests for everything RPC errors/warnings
- Add custom code for `malformedCurrency` and use it in `ledger_entry`

### CODE REVIEW

File: .githooks/pre-commit [hunk 1]
```diff
@@ -6,7 +6,7 @@ exec 1>&2
 find src unittests -type f \( -name '*.cpp' -o -name '*.h' -o -name '*.ipp' \) -print0 | xargs -0 clang-format -i
 
 # check how many lines differ
-lines=$(git diff | wc -l)
+lines=$(git diff src unittests | wc -l)
```
Comment (godexsoft) [inline .githooks/pre-commit:9]: This change was not planned but i realised my mistake from yesterday and decided to fix it here and now

File: src/rpc/Errors.h [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+public:
+    explicit AccountNotFoundError(std::string const& acct) : account(acct)
+    {
+    }
+    const char*
+    what() const throw() override
+    {
+        return account.c_str();
+    }
+};
```
Comment (cjcobb23) [inline src/rpc/Errors.h:146]: Do we actually use these? I know they were here before.

Comment (godexsoft) [reply src/rpc/Errors.h:146]: Yes, seems like we still do use them. We could possibly get rid of them eventually but i'd keep for now to make the changes in this pr minimal (already quite a lot).

File: src/rpc/Errors.h [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+{
+    ClioError const code;
+    std::string_view const error;
+    std::string_view const message;
+};
+
+/**
+ * @brief Clio uses compatible Rippled error codes for most RPC errors.
+ */
+using Error = ripple::error_code_i;
```
Comment (cjcobb23) [inline src/rpc/Errors.h]: Should we call this `RippledError`?

Comment (godexsoft) [reply src/rpc/Errors.h]: We can. I left it as `Error` because it's less changes on call sites - we use this `Error` quite a bit. Let me try to rename it

Comment (cjcobb23) [reply src/rpc/Errors.h]: oh thats true, lots of changes, we could make a separate issue/PR and do that later

Comment (godexsoft) [reply src/rpc/Errors.h]: In the same branch, after changing `Error` to `RippledError` i get `28 files changed, 289 insertions(+), 226 deletions(-)
`. WDYT should we just do it as part of this PR then?

Comment (cjcobb23) [reply src/rpc/Errors.h]: thats not too bad, I'm ok with that if you are, but don't feel strongly

Comment (godexsoft) [reply src/rpc/Errors.h]: Let's just get it done :+1:

**Review by cjcobb23 — APPROVED** [review]
LGTM

**Pull request merged**